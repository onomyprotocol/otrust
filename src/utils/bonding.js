const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
