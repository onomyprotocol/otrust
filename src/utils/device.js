const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
