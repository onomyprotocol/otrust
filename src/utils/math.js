import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
