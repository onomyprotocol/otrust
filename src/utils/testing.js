import { ThemeProvider } from 'styled-components';
import { render } from '@testing-library/react';

import { dark } from 'Theme/theme';

export const renderWithTheme = (Component, rest) => {
  return render(
    <ThemeProvider theme={dark}>
      <Component {...rest} />
    </ThemeProvider>,
  );
};
