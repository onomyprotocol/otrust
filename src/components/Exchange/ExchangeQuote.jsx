import React, { useCallback, useEffect } from 'react';
import { BigNumber } from 'bignumber.js';
import ConfirmTransactionModal from 'components/Modals/components/ConfirmTransactionModal';
import PendingModal from 'components/Modals/components/PendingModal';
import RequestFailedModal from 'components/Modals/components/RequestFailedModal';
import TransactionCompletedModal from 'components/Modals/components/TransactionCompletedModal';
import TransactionFailedModal from 'components/Modals/components/TransactionFailedModal';
import { useWeb3React } from '@web3-react/core';

import { useChain } from 'context/chain/ChainContext';
import { BondingCont } from 'context/chain/contracts';
import { useExchange, useUpdateExchange } from 'context/exchange/ExchangeContext';
import {
  ExchangeItem,
  Sending,
  Receiving,
  ExchangeInput,
  MaxBtn,
  ReceivingValue,
  ExchangeButton,
} from './exchangeStyles';
import { useModal } from 'context/modal/ModalContext';
import NOMButton from 'components/Exchange/NOMButton';
import { format18, parse18 } from 'utils/math';

export default function ExchangeQuote({ strength }) {
  const { strongBalance, weakBalance } = useChain();
  const { handleModal } = useModal();
  const { library } = useWeb3React();

  const bondContract = BondingCont(library);
  // const NOMcontract = NOMCont(library);

  const { askAmount, bidAmount, bidDenom, input, output, strong, weak } = useExchange();

  useEffect(() => {
    console.log('Input: ', input);
    console.log('Output: ', output);
  });

  const { objDispatch, strDispatch } = useUpdateExchange();

  const getAskAmount = useCallback(
    async (askAmountState, bidAmountUpdate, textStrength) => {
      var askAmountUpdate = askAmountState;

      switch (textStrength) {
        case 'strong':
          console.log('Strong: ', bidAmountUpdate.toFixed(0));
          askAmountUpdate = await bondContract.buyQuoteETH(bidAmountUpdate.toFixed(0));
          console.log('Pull Strong Ask Amount', askAmountUpdate);
          break;

        case 'weak':
          askAmountUpdate = await bondContract.sellQuoteNOM(bidAmountUpdate.toFixed(0));
          console.log('Pull Weak Ask Amount', askAmountUpdate);
          break;

        default:
          console.error('Denom not set');
      }
      return new BigNumber(askAmountUpdate.toString());
    },
    [bondContract],
  );

  const onApprove = async () => {
    if (bidAmount <= weakBalance) {
      handleModal(<PendingModal type="approving" />);
      // try {
      //   strDispatch({
      //     type: 'status',
      //     value: 'APPROVE',
      //   });

      //   let tx = await NOMcontract.increaseAllowance(bondContract.address, bidAmount.toFixed(0));

      //   tx.wait().then(() => {
      //     handleModal(<TransactionCompletedModal tx={tx} />);
      //   });

      //   strDispatch({
      //     type: 'status',
      //     value: '',
      //   });
      // } catch (e) {
      //   // eslint-disable-next-line no-console
      //   // console.error(e.code, e.message.message);
      //   // alert(e.message)
      //   handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
      // }
    } else {
      handleModal(<TransactionFailedModal error={`${weak} Balance too low`} />);
    }
  };

  const submitTrans = useCallback(
    async (slippage, gasPrice) => {
      handleModal(<PendingModal />);
      if (!bidAmount || !askAmount) return;
      try {
        let tx;
        let gasFee;

        switch (bidDenom) {
          case 'strong':
            // Preparing for many tokens / coins
            switch (strong) {
              case 'ETH':
                const gasFeeRaw = await bondContract.estimateGas.buyNOM(askAmount.toFixed(0), slippage.toFixed(0), {
                  value: bidAmount.toFixed(0),
                });

                gasFee = new BigNumber(gasFeeRaw.toString());

                const bidAmountUpdate = bidAmount.minus(gasFee.times(gasPrice));
                const askAmountUpdateRaw = await bondContract.buyQuoteETH(bidAmountUpdate.toFixed(0));
                const askAmountUpdate = new BigNumber(askAmountUpdateRaw.toString());

                tx = await bondContract.buyNOM(askAmountUpdate.toFixed(0), slippage.toFixed(0), {
                  value: bidAmountUpdate.toFixed(0),
                  gasPrice: gasPrice.toFixed(0),
                  gasLimit: gasFee.toFixed(0),
                });

                tx.wait().then(() => {
                  handleModal(<TransactionCompletedModal tx={tx} />);
                });

                break;
              default: {
              }
            }
            break;

          case 'weak':
            switch (weak) {
              case 'wNOM':
                tx = await bondContract.sellNOM(bidAmount.toFixed(0), askAmount.toFixed(0), slippage.toFixed(0), {
                  gasPrice: gasPrice.toFixed(0),
                });

                tx.wait().then(() => {
                  handleModal(<TransactionCompletedModal tx={tx} />);
                });
                break;
              default: {
              }
            }
            break;

          default:
            console.log();
        }
      } catch (e) {
        // eslint-disable-next-line no-console
        console.error(e.code, e.message.message);
        // alert(e.message)
        handleModal(<TransactionFailedModal error={e.code + '\n' + e.message.slice(0, 80) + '...'} />);
      }
    },
    [askAmount, bidAmount, bidDenom, bondContract, handleModal, strong, weak],
  );

  const onBid = () => {
    switch (true) {
      case bidDenom !== strength:
        handleModal(<RequestFailedModal error="Please enter amount" />);
        break;
      case strength === 'strong' && strongBalance.gte(bidAmount):
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
        break;
      case strength === 'weak' && weakBalance.gte(bidAmount):
        handleModal(<ConfirmTransactionModal submitTrans={submitTrans} />);
        break;
      default:
        handleModal(<RequestFailedModal error="Insufficient funds" />);
    }
  };

  const onMax = async () => {
    let strUpdate = new Map();
    strUpdate.set('bidDenom', strength);
    let bidMaxValue = strength === 'strong' ? strongBalance : weakBalance;

    strUpdate.set('input', format18(bidMaxValue).toString());

    let askAmountUpdate;

    try {
      askAmountUpdate = await getAskAmount(askAmount, bidMaxValue, strength);
    } catch (err) {
      if (err) {
        handleModal(<RequestFailedModal error={err.error.message} />);
      }
    }

    let objUpdate = new Map();

    objUpdate = objUpdate.set('askAmount', askAmountUpdate);

    objUpdate = objUpdate.set('bidAmount', bidMaxValue);

    objDispatch({
      type: 'update',
      value: objUpdate,
    });

    strUpdate.set('output', format18(new BigNumber(askAmountUpdate.toString())).toFixed(8));

    strDispatch({
      type: 'update',
      value: strUpdate,
    });
  };

  const onTextChange = useCallback(
    async (evt, textStrength) => {
      evt.preventDefault();
      const floatRegExp = new RegExp(/(^(?=.+)(?:[1-9]\d*|0)?(?:\.\d+)?$)|(^\d\.$)/);
      console.log('Component Strength: ', strength);
      console.log('Text Strength: ', textStrength);
      console.log('Bid Denom: ', bidDenom);
      let strUpdate = new Map();
      switch (true) {
        case bidDenom === strength && input === evt.target.value.toString():
          break;
        case evt.target.value === '' || evt.target.value === '.':
          {
            let objUpdate = new Map();

            objUpdate = objUpdate.set('askAmount', new BigNumber(0));

            objUpdate = objUpdate.set('bidAmount', new BigNumber(0));

            objDispatch({
              type: 'update',
              value: objUpdate,
            });
          }

          strUpdate = strUpdate.set('bidDenom', strength);

          strUpdate = strUpdate.set('input', evt.target.value.toString());

          strUpdate = strUpdate.set('output', '');

          strDispatch({
            type: 'update',
            value: strUpdate,
          });

          break;
        case floatRegExp.test(evt.target.value.toString()):
          const evttargetvalue = evt.target.value;

          console.log('Input after test', evttargetvalue);

          const bidAmountUpdate = parse18(new BigNumber(parseFloat(evttargetvalue).toString()));

          const inputUpdate = evttargetvalue.toString();

          if (bidDenom !== strength) {
            strUpdate = strUpdate.set('bidDenom', strength);
          }

          var askAmountUpdate;

          try {
            askAmountUpdate = await getAskAmount(askAmount, bidAmountUpdate, textStrength);
          } catch (err) {
            if (err) {
              console.log(err.error.message);
              handleModal(<RequestFailedModal error={err.error.message} />);
            }
            break;
          }

          let objUpdate = new Map();

          objUpdate = objUpdate.set('askAmount', new BigNumber(askAmountUpdate.toString()));

          objUpdate = objUpdate.set('bidAmount', bidAmountUpdate);

          objDispatch({
            type: 'update',
            value: objUpdate,
          });

          strUpdate = strUpdate.set('input', inputUpdate);

          strUpdate = strUpdate.set('output', format18(new BigNumber(askAmountUpdate.toString())).toFixed(8));
          strDispatch({
            type: 'update',
            value: strUpdate,
          });

          break;
        default:
          handleModal(<RequestFailedModal error="Please enter numbers only. Thank you!" />);
      }
    },
    [askAmount, bidDenom, getAskAmount, handleModal, input, objDispatch, strDispatch, strength],
  );

  return (
    <ExchangeItem>
      <strong>Bid {strength === 'strong' ? strong : weak}</strong>
      <Sending>
        <strong>I'm bidding</strong>
        <ExchangeInput
          type="text"
          onChange={evt => onTextChange(evt, strength)}
          value={bidDenom === strength ? input : ''}
          placeholder="0.00"
        />
        {strength === 'strong' ? strong : weak}
        <MaxBtn onClick={() => onMax()}>Max</MaxBtn>
      </Sending>
      <Receiving>
        <strong>I'm asking</strong>
        <ReceivingValue>
          {strength === bidDenom ? output : ''} {strength === 'strong' ? weak : strong}
        </ReceivingValue>
      </Receiving>
      {strength === 'strong' ? (
        bidDenom === 'weak' ? (
          <ExchangeButton>Input Value</ExchangeButton>
        ) : bidAmount.lte(strongBalance) ? (
          input === '' ? (
            <ExchangeButton>Input Value</ExchangeButton>
          ) : (
            <ExchangeButton onClick={onBid}>Buy {strength === 'strong' ? weak : strong}</ExchangeButton>
          )
        ) : (
          <ExchangeButton>Low {strong} Balance</ExchangeButton>
        )
      ) : (
        <NOMButton onBid={onBid} onApprove={onApprove} />
      )}
    </ExchangeItem>
  );
}
