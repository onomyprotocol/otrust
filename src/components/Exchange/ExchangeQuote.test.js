import { cleanup, fireEvent } from '@testing-library/react';
import { renderWithContext } from 'utils/testing';

import ExchangeQuote from './ExchangeQuote';

describe('Given an ExchangeQuote component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithContext(ExchangeQuote);
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when strong balance changed', () => {
    it('should updated the ask amount', () => {
      const { getByTestId } = renderWithContext(ExchangeQuote);
      const receivingValue = getByTestId('exchange-weak-balance');

      fireEvent.change(getByTestId('exchange-strong-balance-input'), { target: { value: '1' } });

      expect(receivingValue).toHaveTextContent('ETH');
    });
  });

  describe('when max button clicked', () => {
    it('should calculated max balace correctly', () => {
      const { queryByTestId } = renderWithContext(ExchangeQuote);
      const receivingValue = queryByTestId('exchange-weak-balance');

      fireEvent.click(queryByTestId('max-value-button'));

      expect(receivingValue).toHaveTextContent('ETH');
    });
  });
});
