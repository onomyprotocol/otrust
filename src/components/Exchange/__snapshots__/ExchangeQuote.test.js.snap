// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given an ExchangeQuote component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-khIgEk hOVrQw"
  >
    <strong>
      Bid NOM
    </strong>
    <div
      class="sc-hTRkXV buJMih"
    >
      <strong>
        I'm bidding
      </strong>
      <input
        class="sc-laZMeE fEFLDL"
        data-testid="exchange-strong-balance-input"
        placeholder="0.00"
        type="text"
        value=""
      />
      NOM
      <button
        class="sc-gSYDnn jPcBbV"
        data-testid="max-value-button"
      >
        Max
      </button>
    </div>
    <div
      class="sc-jgPyTC bvwgWF"
    >
      <strong>
        I'm asking
      </strong>
      <span
        class="sc-iNiQyp irsaWO"
        data-testid="exchange-weak-balance"
      >
         ETH
      </span>
    </div>
    <button
      class="sc-jffHpj sc-eJocfa fXjZsS colzYt"
    >
       Input Value 
    </button>
  </div>
</DocumentFragment>
`;
